### ISSUE TITLE
ZScript Highlighting

### ISSUE THREAD

**Issue by MajorCooke** [sirjuddington/SLADE#611]
ZScript highlighting can pretty much be done on top of the current DECORATE highlighting as it uses all the same semantics, but more.
It'd be nice if one could add the following words for colors when it comes to ZScript.

- Default

Types:
- Class
- Object
- Vector2
- Vector3
- Name
- String
- Color
- Sound

Properties:

[See here for a full list.](https://github.com/rheit/zdoom/blob/zscript/src/p_mobj.cpp#L157)

--- Colors ---
Deprecated (i.e. x, y, z by themselves) - Dark Red
Read Only (Property) - Dark Grey

And then to have all text files under ZScript/ folder automatically load with it.

That's it. I do not expect things like the ability to find what functions one has, that's a long way off if anything... or at all.

**Comment by Gaerzi** [comment]
It'd make sense IMO to wait a bit more for things to stabilize.

I'm not sure the syntax highlighting used allows to add custom types of colors, so deprecated and read-only stuff would just be like normal properties. Also there's no reason for ZScript to have any deprecated stuff yet. Either stuff is in, and it's in; or it's out, and it's out.

**Comment by MajorCooke** [comment]
I should point out DECORATE in and of itself is entirely deprecated, but all the DECORATE functions EXCEPT for A_ChangeFlag are being supported.

Other than that, fair enough.

**Comment by Gaerzi** [comment]
Let's be clear one moment here because "deprecated" is meaningless.

There is stuff that is SUPPORTED BY THE ENGINE and there is stuff that IS NOT SUPPORTED BY THE ENGINE.

Deprecated, as used in ZDoom means "will remain supported forever but further development beyond that is not expected". ZDoom will still be able to load mods using DECORATE. DECORATE support will be unaffected, it's just that there most likely won't be new features.

ZScript is a different beast. Sure it uses parts of the same backend but the parser at least is different. And the parser is what's of interest here, since syntax highlighting is there just to help write the stuff that gets parsed.

So what does it mean for ZScript to have "x, y, z by themselves" to be deprecated? Either of two things here.

1. You cannot actually use x, y or z by themselves. Then it's not deprecated, it's just not supported. No point in syntax highlighting it.

2. You can use x, y, or z by themselves. It's supported. It's part of the specs. It can be used. It will be used. And ZDoom will have to keep supporting it forever and ever because it will have been used. Then there's no need to give it a different color than other properties do.

TL;DR: if something within ZScript (I'm not talking about within DECORATE here) is to be considered deprecated, now is the time to just remove it from ZScript (I'm still not talking about removing from DECORATE) because otherwise calling it deprecated is just dumb. Deprecated is only for legacy features you wish you would have done differently in the first place, not for stuff you are deliberately adding when there is no need for it to exist at all.

**Comment by sirjuddington** [comment]
For the most part this should be simple enough, but yeah using the built in c++ lexer as it currently does, it's only possible to have 3 different colours for keyword sets (currently used for 'keywords', 'constants' and 'functions'). To support more I'd have to write a custom lexer for wxSTC, or drop it for something else.

**Comment by Gaerzi** [comment]
Also another thing: the lexer doesn't support the period in tokens. So trying to have a "deprecated color" for x is dumb because if it were so, what do you would happen to pos.x ? pos would be the normal property color, . would be black, and x would be deprecated.

**Comment by MajorCooke** [comment]
I was simply going off of what Graf said, is all. He's the one who told me DECORATE would be deprecated BUT... I get your point. He just tends to have an influence with me on the use of the word 'deprecate', heh.

However, I should point out that Graf makes it print warnings for 'using deprecated members'. I.e. x, y, and z. I understand that it's not really possible to change the whole vel.x thing. No matter to me.

Still, it'd be nice to add an entry for ZScript at least so I don't have to keep selecting 'ZDoom Decorate' every time I go to edit my ZScript files.

**Comment by sirjuddington** [comment]
Hmm, with the properties you linked, is it safe to assume that eg. `NAME_VisibleStartAngle` is written "VisibleStartAngle" in zscript?

**Comment by MajorCooke** [comment]
Yup. 

Uh oh. Better tell him of the type then.
